internal/shared_lib_core.go

```diff
@@ -0,0 +1,224 @@
+    send_message_t fn,
+    const uint8_t* msg_ptr,
+    size_t msg_len,
+    uint8_t** out_buf,
+    size_t* out_len,
+    size_t* out_cap
+) {
+    return fn(msg_ptr, msg_len, out_buf, out_len, out_cap);
+}
+
+static inline void call_free_message(
+    free_message_t fn,
+    uint8_t* buf,
+    size_t len,
+    size_t cap
+) {
+    fn(buf, len, cap);
+}
+
+// dlopen wrapper
+static void* open_library(const char* path) {
+    return dlopen(path, RTLD_NOW);
+}
+
+// dlsym wrapper
+static void* load_symbol(void* handle, const char* name) {
+    return dlsym(handle, name);
+}
+
+// dlclose wrapper
+static int close_library(void* handle) {
+    return dlclose(handle);
+}
+*/
+import "C"
+
+type SharedLibCore struct {
+	handle       unsafe.Pointer
+	sendMessage  C.send_message_t
+	freeResponse C.free_message_t
+}
+
+var coreLib *SharedLibCore
+
+// find1PasswordLibPath returns the path to the 1Password shared library
+// (libop_sdk_ipc_client.dylib/.so/.dll) depending on OS.
+func find1PasswordLibPath() (string, error) {
+	hostOS := runtime.GOOS
+
+	core, err := GetExtismCore()
+	if err != nil {
+		return "", err
+	}
+
+	locationsRaw, err := core.Invoke(context.Background(), InvokeConfig{Invocation: Invocation{
+		Parameters: Parameters{
+			MethodName:       "GetDesktopAppIPCClientLocations",
+			SerializedParams: map[string]interface{}{"host_os": hostOS},
+		},
+	}})
+	if err != nil {
+		return "", err
+	}
+
+	var locations []string
+	err = json.Unmarshal([]byte(*locationsRaw), &locations)
+	if err != nil {
+		return "", err
+	}
+
+	for _, libPath := range locations {
+		if _, err := os.Stat(libPath); err == nil {
+			return libPath, nil
+		}
+	}
+
+	return "", fmt.Errorf("1Password desktop application not found")
+}
+
+func GetSharedLibCore() (*CoreWrapper, error) {
+	if coreLib == nil {
+		path, err := find1PasswordLibPath()
+		if err != nil {
+			return nil, err
+		}
+		coreLib, err = loadCore(path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	coreWrapper := CoreWrapper{InnerCore: coreLib}
+
+	return &coreWrapper, nil
+}
+
+func loadCore(path string) (*SharedLibCore, error) {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	handle := C.open_library(cPath)
+	if handle == nil {
+		return nil, errors.New("failed to open library")
+	}
+
+	symbol := C.CString("op_sdk_ipc_send_message")
+	defer C.free(unsafe.Pointer(symbol))
+
+	fnSend := C.load_symbol(handle, symbol)
+	if fnSend == nil {
+		C.close_library(handle)
+		return nil, errors.New("failed to load send_message")
+	}
+
+	symbolFree := C.CString("op_sdk_ipc_free_response")
+	defer C.free(unsafe.Pointer(symbolFree))
+
+	fnFree := C.load_symbol(handle, symbolFree)
+	if fnFree == nil {
+		C.close_library(handle)
+		return nil, errors.New("failed to load free_message")
+	}
+
+	return &SharedLibCore{
+		handle:       handle,
+		sendMessage:  (C.send_message_t)(fnSend),
+		freeResponse: (C.free_message_t)(fnFree),
+	}, nil
+}
+
+func (c *SharedLibCore) Close() {
```

Comment:
I can't find where we're calling this function. Let's make sure this gets called when the shared library needs to be released from memory.

Reply:
Actually, now that I checked what we do for the extism core, I saw that we never release the plugin from memory. It will just stay there for the entire duration of the process, once it's loaded, because it's written in a global variable. So we need to decide between:

1. We remain consistent with Extism and never close the shared lib, once we load it.
2. We implement some sort of reference counting for freeing it, after no client is active anymore. 

What do you think?